### DISCUSSION TITLE
Explain module with hetero graph and edges prediction, not working

### DISCUSSION THREAD

**Discussion by arthurserres** [pyg-team/pytorch_geometric#8714] [Q&A] [answered]
Hello, I am interested in a specific case of the [explain](https://pytorch-geometric.readthedocs.io/en/latest/modules/explain.html) library which I cannot manage to solve.

More precisely, I have a model on HeteroData Graph that is performing edge classification. In particular, my model output is not a unique tensor but rather a 'dictionary' of tensors. (One tensor for each type of data). [On this example](https://pytorch-geometric.readthedocs.io/en/latest/tutorial/explain.html#explaining-node-classification-on-a-heterogeneous-graph) It is mentioned that 'It is assumed that model outputs a single tensor' which is not the case in my example. Indeed, if I try to call the explainer I obtain the error '_dict' object has no attribute 'argmax'_ ...

I would like to know how I can, and I should use the explain library with my settings ? (I already checked the [examples page](https://github.com/pyg-team/pytorch_geometric/tree/master/examples/explain) but it does not help)

**Comment by akihironitta** [comment]
@arthurserres Could I see your script? What does your model look like?

I feel like the example [here](https://github.com/pyg-team/pytorch_geometric/blob/09733b493493083ab7cb09faf60716308e270093/examples/explain/captum_explainer_hetero_link.py#L65-L74) will still be a reasonable reference for your use-case.

**Reply by arthurserres** [reply]
Hi [akihironitta](https://github.com/akihironitta), thanks a lot for the answer. My graph has two different types of edges, let us say _type1_ and _type2_. Therefore, if we compute model(batch.x_dict, batch.edge_index_dict, batch.edge_attr_dict), we obtain a dictionary of tensors where each element of the dictionary corresponds to the prediction of each edge type _type1_ and _type2_. I can provide the exact architecture of the model, but I think that the 'issue' here is that instead of a tensor the output is a dictionary of tensors ( I understood it is quite standard with HeteroData).

For now, the solution I found is to consider a new class of model:

```
class model_trunc(torch.nn.Module):
    def __init__(self, model):
        super().__init__()
        self.model = model
    def forward(self, x_dict, edge_index_dict, edge_attr_dict):
        return self.model(x_dict,edge_index_dict, edge_attr_dict)[('type1')]
```

and to perform the explanation on Model = model_trunc(model). I think it does the job as I am only interested in the explanations of _type1_ and here the output of Model is a single tensor.

Looking back at the example that you provide, I think this is essentially what they do with this EdgeDecoder (They are making sure the output is a single tensor).

Thanks a lot.

Best,

Arthur

**Reply by rusty1s** [reply, marked as answer]
Yes, outputs needs to be a `torch.Tensor`. Your solution works. Alternatively, you can stack the outputs of the two node types together before returning it.
